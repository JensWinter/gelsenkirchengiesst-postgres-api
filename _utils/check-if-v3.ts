export function urlContainsV3(url: string): boolean {
	return url.includes("v3");
}
