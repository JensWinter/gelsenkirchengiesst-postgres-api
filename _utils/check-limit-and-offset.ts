import { VercelRequest, VercelResponse } from "@vercel/node";
import { getEnvs } from "./envs";
const { SUPABASE_MAX_ROWS } = getEnvs();
export function checkLimitAndOffset(
	request: VercelRequest,
	response: VercelResponse
) {
	let { limit: limit_str, offset: offset_str } = <
		{ limit: string; offset: string }
	>request.query;

	if (!limit_str) {
		limit_str = `${SUPABASE_MAX_ROWS}`;
	}
	if (!offset_str) {
		offset_str = "0";
	}
	const limit = parseInt(limit_str, 10);
	const offset = parseInt(offset_str, 10);
	if (isNaN(limit)) {
		return response.status(400).json({ error: "limit needs to be a number" });
	}
	if (isNaN(offset)) {
		return response.status(400).json({ error: "offset needs to be a number" });
	}
	if (limit > SUPABASE_MAX_ROWS) {
		return response.status(400).json({
			error: `limit needs to be smaller than ${SUPABASE_MAX_ROWS}`,
		});
	}
}
