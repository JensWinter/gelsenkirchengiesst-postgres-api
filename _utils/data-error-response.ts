import { PostgrestError } from "@supabase/supabase-js";
import { VercelResponse } from "@vercel/node";

export function checkDataError({
	response,
	data,
	error,
	errorMessage,
}: {
	response: VercelResponse;
	data: unknown | null;
	error: PostgrestError | null;
	errorMessage: string;
}) {
	if (error) {
		return response.status(500).json({ error });
	}

	if (!data) {
		return response.status(500).json({ error: errorMessage });
	}
}
