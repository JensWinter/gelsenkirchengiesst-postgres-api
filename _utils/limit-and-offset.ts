import { VercelRequest, VercelResponse } from "@vercel/node";
import { getEnvs } from "./envs";
const { SUPABASE_MAX_ROWS } = getEnvs();
export function checkLimitAndOffset(
	request: VercelRequest,
	response: VercelResponse
) {
	let { limit: limit_str, offset: offset_str } = <
		{ limit: string; offset: string }
	>request.query;

	if (!limit_str) {
		limit_str = `${SUPABASE_MAX_ROWS}`;
	}
	if (!offset_str) {
		offset_str = "0";
	}
	const limit = parseInt(limit_str, 10);
	const offset = parseInt(offset_str, 10);
	if (isNaN(limit)) {
		return response.status(400).json({ error: "limit needs to be a number" });
	}
	if (isNaN(offset)) {
		return response.status(400).json({ error: "offset needs to be a number" });
	}
	if (limit > SUPABASE_MAX_ROWS) {
		return response.status(400).json({
			error: `limit needs to be smaller than ${SUPABASE_MAX_ROWS}`,
		});
	}
}

export function getLimitAndOffeset(query: {
	[key: string]: string | string[];
}): {
	limit: number;
	offset: number;
} {
	let { limit: limit_str, offset: offset_str } = <
		{ limit: string; offset: string }
	>query;

	if (!limit_str) {
		limit_str = `${SUPABASE_MAX_ROWS}`;
	}
	if (!offset_str) {
		offset_str = "0";
	}
	const limit = parseInt(limit_str, 10);
	const offset = parseInt(offset_str, 10);
	if (isNaN(limit) && isNaN(offset)) {
		return { limit: SUPABASE_MAX_ROWS, offset: 0 };
	}

	if (limit > SUPABASE_MAX_ROWS) {
		return { limit: SUPABASE_MAX_ROWS, offset };
	}

	if (offset < 0) {
		return { limit, offset: 0 };
	}
	return { limit, offset };
}
