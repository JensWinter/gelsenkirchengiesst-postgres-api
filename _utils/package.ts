import fs from "fs";
import path from "path";

export interface Package {
	[key: string]: string | number | Record<string, unknown> | undefined;
	version: string;
	name: string;
	homepage?: string;
	bugs?: {
		url: string;
		// [key: string]: string | number | Record<string, unknown>;
	};
}
export function getPackage(): Package {
	const pkgStr = fs.readFileSync(path.resolve("./package.json"), "utf8");
	const pkg = JSON.parse(pkgStr);
	return pkg;
}
