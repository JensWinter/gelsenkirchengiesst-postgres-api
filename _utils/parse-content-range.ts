import { getEnvs } from "./envs";

const { SUPABASE_SERVICE_ROLE_KEY, SUPABASE_ANON_KEY } = getEnvs();

/**
 * getRange that makes a fetch head request to get the content range. Accepts url and returns a promise with the range object.
 * @example
 * const { error, range } = await getRange("http://localhost:54321/trees",{
 * 	headers:{
 * 		Authorization: `Bearer ${SUPABASE_SERVICE_ROLE_KEY}`,
 * 		apikey: SUPABASE_ANON_KEY,
 * 		"Range-Unit": "items",
 * 		Prefer: "count=exact"
 * 		}
 * 	});
 *
 */
export async function getRange(url: string) {
	let error: Error | null = null;
	let range: ContentRange | null = null;

	const response = await fetch(url, {
		headers: {
			Authorization: `Bearer ${SUPABASE_SERVICE_ROLE_KEY}`,
			apikey: SUPABASE_ANON_KEY,
			"Range-Unit": "items",
			Prefer: "count=exact",
		},
		method: "HEAD",
	});

	if (!response.ok) {
		error = new Error("Response not ok");
	}
	if (response.status === 401) {
		error = new Error("Unauthorized");
	}

	const contentRange = response.headers.get("content-range");
	if (contentRange === null) {
		error = new Error("No content-range header");
	} else {
		const parsedRange = parseContentRange(contentRange);
		if (parsedRange === null) {
			error = new Error("Invalid content-range header");
		} else {
			range = parsedRange;
		}
	}

	return { error, range };
}

/**
 * ContentRange if a value could not be parsed to a number it is set to -1
 */
export interface ContentRange {
	start: number;
	end: number;
	total: number;
}

export function parseContentRange(str: string): ContentRange | null {
	if (!str) {
		return null;
	}
	const arr = str.split("/");
	if (arr.length !== 2) {
		return null;
	}
	const rangeRegex = /^(?<start>\d+|\*)-(?<end>\d+|\*)$/;
	const totalRegex = /^(?<total>\d+|\*|0)$/;
	const rangeMatch = arr[0].match(rangeRegex);
	const totalMatch = arr[1].match(totalRegex);
	const start =
		rangeMatch?.groups?.start === "*"
			? -1
			: parseInt(rangeMatch?.groups?.start ?? "-1");
	const end =
		rangeMatch?.groups?.end === "*"
			? -1
			: parseInt(rangeMatch?.groups?.end ?? "-1");
	const total =
		totalMatch?.groups?.total === "*"
			? -1
			: parseInt(totalMatch?.groups?.total ?? "-1");
	if (isNaN(start) || isNaN(end) || isNaN(total)) {
		return null;
	}

	return { start, end, total };
}
