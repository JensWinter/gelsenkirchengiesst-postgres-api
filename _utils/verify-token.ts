// Verify using getKey callback
// Example uses https://github.com/auth0/node-jwks-rsa as a way to fetch the keys.
import jwksClient, { SigningKey } from "jwks-rsa";
import jwt, {
	VerifyOptions,
	JwtHeader,
	SigningKeyCallback,
} from "jsonwebtoken";
import { getEnvs } from "./envs";
const { JWKS_URI: jwksUri, AUDIENCE: audience, ISSUER: issuer } = getEnvs();
const client = jwksClient({
	jwksUri,
});

export const options: VerifyOptions = {
	audience: audience,
	issuer,
	algorithms: ["RS256"],
};

export function getKey(header: JwtHeader, callback: SigningKeyCallback): void {
	if (!header.kid) {
		throw new Error("Header.kid is missing");
	}

	client.getSigningKey(
		header.kid,

		function (err: Error | null, key?: SigningKey) {
			if (err) {
				callback(err);
				return;
			}
			if (!key) throw new Error("key is missing");

			const signingKey = key.getPublicKey();
			callback(null, signingKey);
		}
	);
}

/**
 * based on https://codedaily.io/tutorials/174/Unit-Test-Token-Verification-for-Auth0-using-Jest-and-mock-jwks
 *
 */
export async function verifyAuth0Token(token: string, options: VerifyOptions) {
	return new Promise((resolve, reject) => {
		jwt.verify(token, getKey, options, (err, decoded) => {
			if (err) {
				reject(err);
				return;
			}
			resolve(decoded);
		});
	});
}
