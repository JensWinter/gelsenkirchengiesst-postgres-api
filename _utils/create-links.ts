import { ContentRange } from "./parse-content-range";

// a function that create links for pagination based limit and offset and range
export function createLinks({
	limit,
	offset,
	range,
	type,
	method,
	requestUrl,
}: {
	limit: number;
	offset: number;
	range: ContentRange | null;
	type: string;
	method: string;
	requestUrl: string;
}) {
	const params = new URLSearchParams(requestUrl.split("?")[1]);
	// remove limit and offset from searchParams since we pass them as arguments
	// and they get some precomputing for SUPABASE_MAX_ROWS
	params.delete("limit");
	params.delete("offset");
	const paramsString =
		params.toString().length > 0 ? "&" + params.toString() : "";
	const links: Record<string, string> = {};
	if (range) {
		// If offset is greater than 0, generate a previous link
		const isPrevLink = offset > 0;
		if (isPrevLink) {
			links.prev = `/${method.toLowerCase()}/${type}?limit=${limit}&offset=${
				offset - limit < 0 ? 0 : offset - limit
			}${params ? paramsString : ""}`;
		}
		// If the offset is less than the total number of items minus 1, generate a next link
		const isNextLink = offset + limit < range.total - 1;
		if (isNextLink) {
			links.next = `/${method.toLowerCase()}/${type}?limit=${limit}&offset=${
				offset + limit
			}${params ? paramsString : ""}`;
		}
	}
	return links;
}
