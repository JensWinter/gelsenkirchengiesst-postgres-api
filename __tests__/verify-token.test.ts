import { test, describe, expect, afterAll, jest } from "@jest/globals";
import { getEnvs } from "../_utils/envs";
/* eslint-disable jest/no-hooks */
import createJWKSMock from "mock-jwks";
import { options, verifyAuth0Token } from "../_utils/verify-token";
jest.mock("../_utils/envs", () => {
	return {
		getEnvs: () => {
			return {
				JWKS_URI: "https://jwks.foo/.well-known/jwks.json",
				AUDIENCE: "foo",
				ISSUER: "foo",
			};
		},
	};
});

const createContext = (url?: string) => {
	const jwks = createJWKSMock(url ? url : "https://jwks.foo/");
	return { jwks };
};
describe("token verification", () => {
	// const { jwks } = createContext();

	// beforeEach(() => {
	//   jwks.start();
	// });

	// afterEach(() => {
	//   jwks.stop();
	// });
	afterAll(() => {
		jest.restoreAllMocks();
	});

	test("should verify a token", async () => {
		const { jwks } = createContext();
		jwks.start();
		const payload = { aud: ["foo"], iss: "foo" };
		const token = jwks.token(payload);
		const data = await verifyAuth0Token(token, options);
		// expect(data).toStrictEqual(payload);

		const { AUDIENCE: audience, ISSUER: issuer } = getEnvs();

		expect(data).toMatchSnapshot({
			aud: [audience],
			iat: expect.any(Number),
			iss: issuer,
		});
		jwks.stop();
	});

	test("should thorw errors", async () => {
		const { jwks } = createContext();
		jwks.start();
		const payload = { aud: [], iss: "foo" };
		const token = jwks.token(payload);
		await expect(verifyAuth0Token(token, options)).rejects.toThrow(
			"jwt audience invalid. expected: foo"
		);
		jwks.stop();
	});

	test("should thorw an error", async () => {
		const { jwks } = createContext("https://other.xyz");
		jwks.start();
		const payload = { aud: [], iss: "foo" };
		const token = jwks.token(payload);
		jwks.stop();
		await expect(verifyAuth0Token(token, options)).rejects.toThrow();
	});
});
