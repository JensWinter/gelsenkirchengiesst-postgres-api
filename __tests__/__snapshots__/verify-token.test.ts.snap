// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`token verification should verify a token 1`] = `
{
  "aud": [
    "foo",
  ],
  "iat": Any<Number>,
  "iss": "foo",
}
`;
